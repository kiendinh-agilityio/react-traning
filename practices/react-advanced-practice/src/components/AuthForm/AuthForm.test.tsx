import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { useNavigate } from 'react-router-dom';
import AuthForm from '.';
import { MESSAGE_ERROR } from '@/constants';

jest.mock('@/stores', () => ({
  useAuthStore: jest.fn(() => ({
    showPassword: false,
    togglePasswordVisibility: jest.fn(),
    setName: jest.fn(),
    setEmail: jest.fn(),
    setPassword: jest.fn(),
  })),
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');

describe('AuthForm Component', () => {
  const mockOnSubmit = jest.fn();
  const mockNavigate = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
  });

  const renderComponent = (type: 'signin' | 'signup' = 'signin') =>
    render(
      <BrowserRouter>
        <AuthForm
          type={type}
          onSubmit={mockOnSubmit}
          bottomText="Don't have an account?"
          bottomLink="/signup"
          bottomLinkText="Sign up"
        />
      </BrowserRouter>,
    );

  it('matches the snapshot for signin', () => {
    const { asFragment } = renderComponent('signin');
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches the snapshot for signup', () => {
    const { asFragment } = renderComponent('signup');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders email and password fields', () => {
    renderComponent();
    expect(screen.getByPlaceholderText('Your email address')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Your password')).toBeInTheDocument();
  });

  it('renders name field for signup', () => {
    renderComponent('signup');
    expect(screen.getByPlaceholderText('Your full name')).toBeInTheDocument();
  });

  it('shows an error when submitting an empty form', async () => {
    const { asFragment } = renderComponent();
    fireEvent.click(screen.getByRole('button', { name: /sign in/i }));
    await waitFor(() => {
      expect(screen.getByText(MESSAGE_ERROR.REQUIRED_ERROR('Email'))).toBeInTheDocument();
      expect(
        screen.getByText(MESSAGE_ERROR.REQUIRED_ERROR('Password')),
      ).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('validates email format', async () => {
    renderComponent();
    const emailInput = screen.getByPlaceholderText('Your email address');
    fireEvent.change(emailInput, { target: { value: 'invalid email' } });
    fireEvent.blur(emailInput);
    await waitFor(() => {
      expect(screen.getByText(MESSAGE_ERROR.INVALID_EMAIL)).toBeInTheDocument();
    });
  });

  it('validates password format', async () => {
    renderComponent();
    const passwordInput = screen.getByPlaceholderText('Your password');
    fireEvent.change(passwordInput, { target: { value: '123' } });
    fireEvent.blur(passwordInput);
    await waitFor(() => {
      expect(screen.getByText(MESSAGE_ERROR.INVALID_PASSWORD)).toBeInTheDocument();
    });
  });

  it('shows an error when name is empty in signup', async () => {
    renderComponent('signup');
    fireEvent.click(screen.getByRole('button', { name: /sign up/i }));
    await waitFor(() => {
      expect(screen.getByText(MESSAGE_ERROR.REQUIRED_ERROR('Name'))).toBeInTheDocument();
    });
  });

  it('calls onSubmit with valid data on signin', async () => {
    const { asFragment } = renderComponent();

    fireEvent.input(screen.getByPlaceholderText('Your email address'), {
      target: { value: 'test@example.com' },
    });
    fireEvent.input(screen.getByPlaceholderText('Your password'), {
      target: { value: 'Password@123' },
    });

    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(screen.getByRole('button', { name: /sign in/i }));

    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalledWith(
        expect.objectContaining({
          email: 'test@example.com',
          password: 'Password@123',
        }),
      );
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls onSubmit with valid data on signup', async () => {
    renderComponent('signup');

    fireEvent.input(screen.getByPlaceholderText('Your full name'), {
      target: { value: 'John Doe' },
    });
    fireEvent.input(screen.getByPlaceholderText('Your email address'), {
      target: { value: 'john@example.com' },
    });
    fireEvent.input(screen.getByPlaceholderText('Your password'), {
      target: { value: 'Password@123' },
    });
    fireEvent.click(screen.getByRole('button', { name: /sign up/i }));

    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalledWith({
        name: 'John Doe',
        email: 'john@example.com',
        password: 'Password@123',
      });
    });
  });

  it('navigates to /home after successful submission', async () => {
    mockOnSubmit.mockResolvedValueOnce(Promise.resolve());

    renderComponent();

    fireEvent.input(screen.getByPlaceholderText('Your email address'), {
      target: { value: 'test@example.com' },
    });
    fireEvent.input(screen.getByPlaceholderText('Your password'), {
      target: { value: 'Password@123' },
    });
    fireEvent.click(screen.getByRole('button', { name: /sign in/i }));

    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalledTimes(1);
    });

    jest.advanceTimersByTime(1500);

    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith('/home');
    });
  });

  it('shows error message when submission fails', async () => {
    mockOnSubmit.mockRejectedValueOnce(new Error('Login failed'));
    renderComponent();

    fireEvent.input(screen.getByPlaceholderText('Your email address'), {
      target: { value: 'test@example.com' },
    });
    fireEvent.input(screen.getByPlaceholderText('Your password'), {
      target: { value: 'Password@123' },
    });
    fireEvent.click(screen.getByRole('button', { name: /sign in/i }));

    await waitFor(() => {
      expect(screen.getByText(MESSAGE_ERROR.INVALID_SIGNIN)).toBeInTheDocument();
    });
  });
});
