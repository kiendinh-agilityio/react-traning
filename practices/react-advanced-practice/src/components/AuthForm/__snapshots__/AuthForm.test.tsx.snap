// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm Component matches the snapshot for signin 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col max-w-[350px]"
  >
    <div
      class="rt-Flex rt-r-fd-column mb-6 gap-[5px]"
    >
      <p
        class="rt-Text rt-r-weight-regular text-sm "
      >
        Email
      </p>
      <div
        class="rt-Box w-full"
      >
        <div
          class="rt-Flex rt-r-ai-center rt-r-jc-center flex justify-center items-center px-5 py-[15px] border gap-3 w-full rounded-[15px] border-input "
        >
          <input
            class="w-full text-regular focus:outline-none text-dark placeholder:text-base dark:bg-dark "
            name="email"
            placeholder="Your email address"
            type="email"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="rt-Flex rt-r-fd-column gap-[5px]"
    >
      <p
        class="rt-Text rt-r-weight-regular text-sm "
      >
        Password
      </p>
      <div
        class="rt-Box w-full"
      >
        <div
          class="rt-Flex rt-r-ai-center rt-r-jc-center flex justify-center items-center px-5 py-[15px] border gap-3 w-full rounded-[15px] border-input "
        >
          <input
            class="w-full text-regular focus:outline-none text-dark placeholder:text-base dark:bg-dark "
            name="password"
            placeholder="Your password"
            type="password"
            value=""
          />
          <button
            aria-label="Show password"
            type="button"
          >
            <svg
              fill="none"
              height="23"
              viewBox="0 0 28 28"
              width="23"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                clip-rule="evenodd"
                fill="#a0aec0"
                fill-rule="evenodd"
              >
                <path
                  d="M17.747 15.415c.238-.54.372-1.143.372-1.413 0-2.348-1.824-4.25-4.073-4.25-2.25 0-4.073 1.902-4.073 4.25 0 2.347 2.037 3.887 4.073 3.887 1.303 0 2.462-.637 3.206-1.63a4.29 4.29 0 0 0 .495-.844Zm-3.701.349c.51 0 .975-.196 1.332-.518.432-.39.622-.965.622-1.244 0-1.174-.828-2.126-1.954-2.126-1.125 0-2.046.95-2.046 2.124 0 .81.92 1.764 2.046 1.764Z"
                />
                <path
                  d="M1.092 14.272a.448.448 0 0 1 0-.545c.01-.012.019-.025.028-.039C4.78 8.343 9.32 5.5 14.014 5.5c4.689 0 9.224 2.836 12.882 8.17.069.101.104.205.104.33 0 .13-.04.24-.114.346C23.23 19.67 18.698 22.5 14.014 22.5c-4.705 0-9.257-2.857-12.922-8.228Zm2.847-.92a1.04 1.04 0 0 0 0 1.294c3.135 3.898 6.665 5.729 10.075 5.729 3.41 0 6.94-1.83 10.075-5.729a1.04 1.04 0 0 0 0-1.294c-3.135-3.897-6.665-5.727-10.075-5.727-3.41 0-6.94 1.83-10.075 5.727Z"
                />
              </g>
            </svg>
          </button>
        </div>
      </div>
    </div>
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button  
    px-[22px] py-2.5 font-bold rounded-xl cursor-pointer primary
   mt-9 w-[350px]"
      data-accent-color=""
    >
      Sign In
    </button>
    <div
      class="rt-Text rt-r-weight-regular text-sm font-regular text-base text-center mt-[22px]"
    >
      Don't have an account? 
      <a
        class="rt-Text rt-reset rt-Link rt-underline-auto cursor-pointer text-primary"
        data-accent-color=""
        href="/signup"
        target="_self"
      >
        Sign up
      </a>
    </div>
  </form>
</DocumentFragment>
`;

exports[`AuthForm Component matches the snapshot for signup 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col max-w-[350px]"
  >
    <div
      class="rt-Flex rt-r-fd-column mb-6 gap-[5px]"
    >
      <p
        class="rt-Text rt-r-weight-regular text-sm "
      >
        Name
      </p>
      <div
        class="rt-Box w-full"
      >
        <div
          class="rt-Flex rt-r-ai-center rt-r-jc-center flex justify-center items-center px-5 py-[15px] border gap-3 w-full rounded-[15px] border-input "
        >
          <input
            class="w-full text-regular focus:outline-none text-dark placeholder:text-base dark:bg-dark "
            name="name"
            placeholder="Your full name"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="rt-Flex rt-r-fd-column mb-6 gap-[5px]"
    >
      <p
        class="rt-Text rt-r-weight-regular text-sm "
      >
        Email
      </p>
      <div
        class="rt-Box w-full"
      >
        <div
          class="rt-Flex rt-r-ai-center rt-r-jc-center flex justify-center items-center px-5 py-[15px] border gap-3 w-full rounded-[15px] border-input "
        >
          <input
            class="w-full text-regular focus:outline-none text-dark placeholder:text-base dark:bg-dark "
            name="email"
            placeholder="Your email address"
            type="email"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="rt-Flex rt-r-fd-column gap-[5px]"
    >
      <p
        class="rt-Text rt-r-weight-regular text-sm "
      >
        Password
      </p>
      <div
        class="rt-Box w-full"
      >
        <div
          class="rt-Flex rt-r-ai-center rt-r-jc-center flex justify-center items-center px-5 py-[15px] border gap-3 w-full rounded-[15px] border-input "
        >
          <input
            class="w-full text-regular focus:outline-none text-dark placeholder:text-base dark:bg-dark "
            name="password"
            placeholder="Your password"
            type="password"
            value=""
          />
          <button
            aria-label="Show password"
            type="button"
          >
            <svg
              fill="none"
              height="23"
              viewBox="0 0 28 28"
              width="23"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                clip-rule="evenodd"
                fill="#a0aec0"
                fill-rule="evenodd"
              >
                <path
                  d="M17.747 15.415c.238-.54.372-1.143.372-1.413 0-2.348-1.824-4.25-4.073-4.25-2.25 0-4.073 1.902-4.073 4.25 0 2.347 2.037 3.887 4.073 3.887 1.303 0 2.462-.637 3.206-1.63a4.29 4.29 0 0 0 .495-.844Zm-3.701.349c.51 0 .975-.196 1.332-.518.432-.39.622-.965.622-1.244 0-1.174-.828-2.126-1.954-2.126-1.125 0-2.046.95-2.046 2.124 0 .81.92 1.764 2.046 1.764Z"
                />
                <path
                  d="M1.092 14.272a.448.448 0 0 1 0-.545c.01-.012.019-.025.028-.039C4.78 8.343 9.32 5.5 14.014 5.5c4.689 0 9.224 2.836 12.882 8.17.069.101.104.205.104.33 0 .13-.04.24-.114.346C23.23 19.67 18.698 22.5 14.014 22.5c-4.705 0-9.257-2.857-12.922-8.228Zm2.847-.92a1.04 1.04 0 0 0 0 1.294c3.135 3.898 6.665 5.729 10.075 5.729 3.41 0 6.94-1.83 10.075-5.729a1.04 1.04 0 0 0 0-1.294c-3.135-3.897-6.665-5.727-10.075-5.727-3.41 0-6.94 1.83-10.075 5.727Z"
                />
              </g>
            </svg>
          </button>
        </div>
      </div>
    </div>
    <button
      class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button  
    px-[22px] py-2.5 font-bold rounded-xl cursor-pointer primary
   mt-9 w-[350px]"
      data-accent-color=""
    >
      Sign Up
    </button>
    <div
      class="rt-Text rt-r-weight-regular text-sm font-regular text-base text-center mt-[22px]"
    >
      Don't have an account? 
      <a
        class="rt-Text rt-reset rt-Link rt-underline-auto cursor-pointer text-primary"
        data-accent-color=""
        href="/signup"
        target="_self"
      >
        Sign up
      </a>
    </div>
  </form>
</DocumentFragment>
`;
